### PR TITLE
Faster encode technique

### PR DESCRIPTION
**Pull request by cantino**
(I've learned that there's often a reason that something is written the way it is, so please tell me if I'm overlooking something critical with this change.)

This PR switches the `encode` method to use a combined Regexp with block substitution instead of doing the per-codepoint iteration.  I found on my test runs that switching to a Regexp resulted in a very significant speed up.  Additionally, I've added some memoization and caching for the older method as well because it's currently something like O(C * E * A) where C is the number of codepoints, E is the number of Emoji, and A is the average number of codepoints per Emoji.

Feedback very much appreciated!

### CODE REVIEW

**Comment by mwunsch** [comment]
This looks fantastic, but I would love to be able to take advantage of this in `encode_io`. I want to support text streams as first class. Maybe we can scan the input line-by-line and `gsub` instead of scanning the text codepoint-by-codepoint. Thoughts?

**Comment by cantino** [comment]
Hey @mwunsch.  I've been running into issues with the encode_io implementation, so I'd be supportive of changing it.  (Actually, I [filed a bug in Ruby](https://bugs.ruby-lang.org/issues/11444) around the hangs that I've been getting and it looks like they fixed it-- in 24 hours no less!)

If we break it into lines, how will we know if we're in the middle of a multi-codepoint emoji in the stream?  I could imagine either coming up with a split regex that never breaks inside of what could be a multi-codepoint emoji, or maybe just splitting on new line would work?

**Comment by cantino** [comment]
Okay, I've implemented `encode_io` in terms of `encode`.  It's *much* faster than before.  Can you think of any issues with this approach?

File: bin/rumoji [hunk 1]
```diff
@@ -1,6 +1,13 @@
 #!/usr/bin/env ruby
 # -*- encoding: utf-8 -*-
 
+$: << File.expand_path(File.join(File.dirname(__FILE__), '..', 'lib'))
+
+require 'rubygems'
```
Comment (mwunsch) [inline bin/rumoji]: These two lines have to go. This bin is designed to be run through `bundle exec` and that should set up the load path properly.

Comment (cantino) [inline bin/rumoji]: Done.

**Comment by mwunsch** [comment]
:+1:

**Comment by cantino** [comment]
Great! :)

File: lib/rumoji/emoji/symbols.rb [hunk 2]
```diff
@@ -75,7 +75,7 @@ class Emoji
       self.new("\u{1f564}", [:clock930]),
       self.new("\u{3297}" , [:congratulations]),
       self.new("\u{1f192}", [:cool]),
-      self.new("\u{00a9}" , [:copyright]),
```
Comment (cantino) [inline lib/rumoji/emoji/symbols.rb]: I assume you noted that I commented these out?  We didn't like it replacing normal, non-emoji characters.

**Pull request merged**